node/bindnode/infer.go

```diff
@@ -87,6 +104,78 @@ func fieldNameFromSchema(name string) string {
 	return strings.Title(name)
 }
 
+var defaultTypeSystem schema.TypeSystem
+
+func init() {
+	defaultTypeSystem.Init()
+
+	defaultTypeSystem.Accumulate(schemaTypeBool)
+	defaultTypeSystem.Accumulate(schemaTypeInt)
+	defaultTypeSystem.Accumulate(schemaTypeFloat)
+	defaultTypeSystem.Accumulate(schemaTypeString)
+	defaultTypeSystem.Accumulate(schemaTypeBytes)
+	defaultTypeSystem.Accumulate(schemaTypeLink)
+}
+
+// TODO: support IPLD maps and unions in inferSchema
+
+// TODO: support bringing your own TypeSystem?
```

Comment:
this can probably happen in a subsequent iteration

Reply:
Yep indeed. Just trying to keep PRs small :)

Reply:
Here's one reason why having a global default can be a bad idea: if package A wraps its named type T1, that gets registered in the global typesystem. If package B wraps its own different named type T1, then the two names conflict.

I don't want to use one typesystem per Wrap call, because that's just wasteful.

I don't want to force the user to supply a TypeSystem, because most people will just not want to care about this implementation detail.

But one single global fallback is also not enough, because the names of Go types are package-scoped, not universe-scoped.

Maybe one fallback TypeSystem per Go package would be the sweet spot. A similar alternative would be to use fully qualified Go names as schema type names, like `A.T1` instead of just `T1`, though I'm not sure if that's not okay in IPLD land.

Reply:
My initial instinct is that a munged version of the fully qualified go name is the right way to register these reflected types. It may need to be `A_T1` rather than `A.T1`, but that seems to have the correct level of disambiguation that's desired.

Reply:
I actually think that one typesystem per Go package just won't work, because it's reasonable to want to call bindnode.Wrap on A.T1 which references B.T2. And the wrapped B.T2 must be in the same typesystem for IPLD to work.

I agree that it seems like "fully qualified names" seems the best approach, be it `pkg/path_T1`, `pkg_path_T1`, `pkg/path.T1`, etc.

Reply:
> I don't want to force the user to supply a TypeSystem, because most people will just not want to care about this implementation detail.

I think of all the constraints we could choose to drop here: I'd probably pick this one?

How would it feel to have a `bindnode.Analyze(goValue interface{}) (schema.TypeSystem, error)` API?

Reply:
On second thought: the concept of an `Analyze` call like that doesn't necessarily solve the problem, either, because it's possible to have a type system that has more than one "root".  (It's a full directed graph; it's not guaranteed to have a single start node from which a walk will reach all other nodes.)  That seems to point to a more fundamental challenge: either the analysis phase has to take multiple parameters (not easy to use), or we'd have to have mutable `schema.TypeSystem` (which... oof, interesting).

Reply:
Maybe the sweet spot actually is creating one implicit typesystem per Wrap call, and nevermind the waste, because anything else seems to be either complicated or have potential to generate user surprise.

If the typesystem inference path is only for quick-and-dirty rapid development anyway, it seems fine if it has poor performance.

Reply:
We could always have our cake and eat it too: add `...Option` to the bindnode APIs, and have one option where the user can supply a typesystem. If none is supplied, a global fallback is used. I think that would be good enough for most people: those who want to be in charge of their typesystems can do that, and those who don't know what that means will probably be fine with using a global one (with fully qualified type names to avoid name clashes).

I'm not sure what you mean by mutable TypeSystem. For a global one, we'd simply accumulate more types as needed. I think the typesystem supports accumulating some types, validating the graph, using that typesystem, and later on repeating that process if we infer more IPLD schemas.

Reply:
> If the typesystem inference path is only for quick-and-dirty rapid development anyway, it seems fine if it has poor performance.

The current implementation is certainly quick-and-dirty. I'm leaving performance for later, for the sake of getting things working first. But I definitely don't intend schema inferrence as only a temporary step with poor performance; one should be able to get it to a point where it's fast enough for most non-bottleneck use cases. Deduplicating work is certainly part of that, just like how encoding/json does with a global `sync.Map` to avoid redoing reflect work :)